src/main/java/org/datadog/jenkins/plugins/datadog/tracer/DatadogTracerConfigurator.java

```diff
@@ -0,0 +1,175 @@
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.datadog.jenkins.plugins.datadog.DatadogClient;
+import org.datadog.jenkins.plugins.datadog.DatadogGlobalConfiguration;
+import org.datadog.jenkins.plugins.datadog.DatadogUtilities;
+
+public class DatadogTracerConfigurator {
+
+    private static final Logger LOGGER = Logger.getLogger(DatadogTracerConfigurator.class.getName());
+
+    private final Map<TracerLanguage, TracerConfigurator> configurators;
+
+    public static final DatadogTracerConfigurator INSTANCE = new DatadogTracerConfigurator();
+
+    public DatadogTracerConfigurator() {
+        configurators = new EnumMap<>(TracerLanguage.class);
+        configurators.put(TracerLanguage.JAVA, new JavaConfigurator());
+        configurators.put(TracerLanguage.JAVASCRIPT, new JavascriptConfigurator());
+        configurators.put(TracerLanguage.PYTHON, new PythonConfigurator());
+    }
+
+    public Map<String, String> configure(Run<?, ?> run, Node node, Map<String, String> envs) {
+        Job<?, ?> job = run.getParent();
+        DatadogTracerJobProperty<?> tracerConfig = job.getProperty(DatadogTracerJobProperty.class);
+        if (tracerConfig == null || !tracerConfig.isOn()) {
+            return Collections.emptyMap();
+        }
+
+        Collection<TracerLanguage> languages = tracerConfig.getLanguages();
+        for (ConfigureTracerAction action : run.getActions(ConfigureTracerAction.class)) {
+            if (action.node == node && languages.equals(action.languages)) {
+                return action.variables;
+            }
+        }
+
+        DatadogGlobalConfiguration datadogConfig = DatadogUtilities.getDatadogGlobalDescriptor();
+        if (datadogConfig == null) {
+            LOGGER.log(Level.INFO, "Cannot set up tracer: Datadog config not found");
+            return Collections.emptyMap();
+        }
+
+        TopLevelItem topLevelItem = getTopLevelItem(run);
+        FilePath workspacePath = node.getWorkspaceFor(topLevelItem);
+        if (workspacePath == null) {
+            throw new IllegalStateException("Cannot find workspace path for " + topLevelItem + " on " + node);
+        }
+
+        Map<String, String> variables = new HashMap<>(getCommonEnvVariables(datadogConfig, tracerConfig));
+        for (TracerLanguage language : languages) {
+            TracerConfigurator tracerConfigurator = configurators.get(language);
+            if (tracerConfigurator == null) {
+                LOGGER.log(Level.INFO, "Cannot find tracer configurator for " + language);
+                continue;
+            }
+
+            try {
+                Map<String, String> languageVariables = tracerConfigurator.configure(tracerConfig, node, workspacePath, envs);
+                variables.putAll(languageVariables);
+            } catch (Exception e) {
+                LOGGER.log(Level.INFO, "Error while configuring " + language + " Datadog Tracer for run " + run + " and node " + node, e);
```

Comment:
I think this should be a higher severity log

Reply:
Makes sense, increased severity

Reply:
Thanks!